File: multi_x_serverless/routing/deployment_input/input_manager.py
import random
import time
from typing import Optional

from multi_x_serverless.common.constants import TAIL_LATENCY_THRESHOLD
from multi_x_serverless.common.models.endpoints import Endpoints
from multi_x_serverless.common.models.remote_client.remote_client import RemoteClient
from multi_x_serverless.routing.deployment_input.components.calculators.carbon_calculator import CarbonCalculator
from multi_x_serverless.routing.deployment_input.components.calculators.cost_calculator import CostCalculator
from multi_x_serverless.routing.deployment_input.components.calculators.runtime_calculator import RuntimeCalculator
from multi_x_serverless.routing.deployment_input.components.loaders.carbon_loader import CarbonLoader
from multi_x_serverless.routing.deployment_input.components.loaders.datacenter_loader import DatacenterLoader
from multi_x_serverless.routing.deployment_input.components.loaders.performance_loader import PerformanceLoader
from multi_x_serverless.routing.deployment_input.components.loaders.region_viability_loader import RegionViabilityLoader
from multi_x_serverless.routing.deployment_input.components.loaders.workflow_loader import WorkflowLoader
from multi_x_serverless.routing.models.instance_indexer import InstanceIndexer
from multi_x_serverless.routing.models.region_indexer import RegionIndexer
from multi_x_serverless.routing.workflow_config import WorkflowConfig

random.seed(time.time())


class InputManager:  # pylint: disable=too-many-instance-attributes
    _region_indexer: RegionIndexer
    _instance_indexer: InstanceIndexer
    _execution_latency_distribution_cache: dict[str, list[float]]
    _invocation_probability_cache: dict[str, float]

    def __init__(self, workflow_config: WorkflowConfig, tail_latency_threshold: int = TAIL_LATENCY_THRESHOLD) -> None:
        super().__init__()
        # Set the workflow config
        self._workflow_config: WorkflowConfig = workflow_config

        # Save the tail threshold (Number = percentile for tail runtime/latency)
        # This value MUST be between 50 and 100
        if tail_latency_threshold < 50 or tail_latency_threshold > 100:
            raise ValueError("Tail threshold must be between 50 and 100")
        self._tail_latency_threshold: float = tail_latency_threshold

        # Initialize remote client
        self._data_collector_client: RemoteClient = Endpoints().get_data_collector_client()

        # Initialize loaders
        self._region_viability_loader = RegionViabilityLoader(self._data_collector_client)
        self._datacenter_loader = DatacenterLoader(self._data_collector_client)
        self._performance_loader = PerformanceLoader(self._data_collector_client)
        self._carbon_loader = CarbonLoader(self._data_collector_client)
        self._workflow_loader = WorkflowLoader(self._data_collector_client, workflow_config)

        # Setup the viability loader and load the availability regions
        self._region_viability_loader.setup()  # Setup the viability loader -> This loads data from the database

        # Setup the calculator
        self._runtime_calculator = RuntimeCalculator(self._performance_loader, self._workflow_loader)
        self._carbon_calculator = CarbonCalculator(
            self._carbon_loader, self._datacenter_loader, self._workflow_loader, self._runtime_calculator
        )
        self._cost_calculator = CostCalculator(self._datacenter_loader, self._workflow_loader, self._runtime_calculator)

    def setup(self, regions_indexer: RegionIndexer, instance_indexer: InstanceIndexer) -> None:
        self._region_indexer = regions_indexer
        self._instance_indexer = instance_indexer

        # Get a set of all chosen regions (regions inside the region_indexer)
        requested_regions: set[str] = set(regions_indexer.get_value_indices().keys())

        # Load the workflow loader
        workflow_id = self._workflow_config.workflow_id
        if workflow_id is None:
            raise ValueError("Workflow ID is not set in the config")
        self._workflow_loader.setup(workflow_id)

        # Home region of the workflow should already be in requested regions
        # We should asset this is true
        if self._workflow_loader.get_home_region() not in requested_regions:
            raise ValueError("Home region of the workflow is not in the requested regions! This should NEVER happen!")

        # Now setup all appropriate loaders
        self._datacenter_loader.setup(requested_regions)
        self._performance_loader.setup(requested_regions)
        self._carbon_loader.setup(requested_regions)

        # Clear cache
        self._invocation_probability_cache: dict[str, float] = {}
        self._execution_latency_distribution_cache: dict[str, list[float]] = {}

    def get_execution_cost_carbon_latency(self, instance_index: int, region_index: int) -> tuple[float, float, float]:
        # Convert the instance and region indices to their names
        instance_name = self._instance_indexer.index_to_value(instance_index)
        region_name = self._region_indexer.index_to_value(region_index)

        # For this we would need the excecution latency.
        ## First check if the value is already in the cache
        key = f"{instance_index}_{region_index}"
        if key in self._execution_latency_distribution_cache:
            execution_latency_distribution = self._execution_latency_distribution_cache[key]
        else:
            # If not, calculate the value and store it in the cache
            execution_latency_distribution = self._runtime_calculator.calculate_runtime_distribution(
                instance_name, region_name
            )
            self._execution_latency_distribution_cache[key] = execution_latency_distribution

        # Now we can get a random sample from the distribution
        execution_latency = execution_latency_distribution[
            int(random.random() * (len(execution_latency_distribution) - 1))
        ]

        # Now we can calculate the cost and carbon
        cost = self._cost_calculator.calculate_execution_cost(instance_name, region_name, execution_latency)
        carbon = self._carbon_calculator.calculate_execution_carbon(instance_name, region_name, execution_latency)

        return (cost, carbon, execution_latency)

    def get_transmission_cost_carbon_latency(
        self,
        from_instance_index: int,
        to_instance_index: int,
        from_region_index: int,
        to_region_index: int,
    ) -> tuple[float, float, float]:
        # Convert the instance and region indices to their names
        from_instance_name: Optional[str] = None
        if from_instance_index != -1:
            from_instance_name = self._instance_indexer.index_to_value(from_instance_index)
        to_instance_name = self._instance_indexer.index_to_value(to_instance_index)
        from_region_name = self._region_indexer.index_to_value(from_region_index)
        to_region_name = self._region_indexer.index_to_value(to_region_index)

        # Get the transmission size distribution
        transmission_size_distribution: list[float] = self._runtime_calculator.get_transmission_size_distribution(
            from_instance_name, to_instance_name, from_region_name, to_region_name
        )

        # Pick a transmission size or default to None
        transmission_size: Optional[float] = None
        if len(transmission_size_distribution) > 0:
            transmission_size = transmission_size_distribution[
                int(random.random() * (len(transmission_size_distribution) - 1))
            ]

        # Get the transmission latency distribution
        transmission_latency_distribution: list[float] = self._runtime_calculator.get_transmission_latency_distribution(
            from_instance_name, to_instance_name, from_region_name, to_region_name, transmission_size
        )

        # Now we can get a random sample from the distribution
        transmission_latency: float = transmission_latency_distribution[
            int(random.random() * (len(transmission_latency_distribution) - 1))
        ]

        if (
            transmission_size is None
        ):  # At this point we can assume that the transmission size is 0 for any missing data
            transmission_size = 0.0

        # Now we can calculate the cost and carbon
        # If start hop, no transmission cost
        cost = 0.0
        if from_instance_name is not None:  # Start hop should not incur transmission cost
            cost = self._cost_calculator.calculate_transmission_cost(
                from_region_name,
                to_region_name,
                transmission_size,
            )

        carbon = self._carbon_calculator.calculate_transmission_carbon(
            from_region_name, to_region_name, transmission_size
        )

        return (cost, carbon, transmission_latency)

    def alter_carbon_setting(self, carbon_setting: Optional[str]) -> None:
        """
        Input should either be 'None' or a string from '0' to '23' indicating the hour of the day.
        """
        self._carbon_calculator.alter_carbon_setting(carbon_setting)
        self._runtime_calculator.reset_cache()

        # Clear the cache
        self._execution_latency_distribution_cache = {}
        self._invocation_probability_cache = {}

    def get_invocation_probability(self, from_instance_index: int, to_instance_index: int) -> float:
        """
        Return the probability of the edge being triggered.
        """
        # Check if the value is already in the cache
        key = f"{from_instance_index}_{to_instance_index}"
        if key in self._invocation_probability_cache:
            return self._invocation_probability_cache[key]

        # Convert the instance indices to their names
        from_instance_name = self._instance_indexer.index_to_value(from_instance_index)
        to_instance_name = self._instance_indexer.index_to_value(to_instance_index)

        # If not, retrieve the value from the workflow loader
        invocation_probability = self._workflow_loader.get_invocation_probability(from_instance_name, to_instance_name)
        self._invocation_probability_cache[key] = invocation_probability
        return invocation_probability

    def get_all_regions(self) -> list[str]:
        return self._region_viability_loader.get_available_regions()

File: caribou-go/src/deployment-input/inputmanager.go
package deployment_input

import (
	"fmt"
	"log"
	"math"
	"sort"
	"strconv"
	"strings"

	"caribou-go/src/deployment-input/components/calculators"
	"caribou-go/src/deployment-input/components/loaders"
	"caribou-go/src/models"
)

type InputManager struct {
	RegionViabilityLoader      *loaders.RegionViabilityLoader
	DataCenterLoader           *loaders.DataCenterLoader
	PerformanceLoader          *loaders.PerformanceLoader
	CarbonLoader               *loaders.CarbonLoader
	WorkflowLoader             *loaders.WorkflowLoader
	RuntimeCalculator          *calculators.RuntimeCalculator
	CarbonCalculator           *calculators.CarbonCalculator
	CostCalculator             *calculators.CostCalculator
	InstanceIndexer            *models.InstanceIndexer
	RegionIndexer              *models.RegionIndexer
	executionLatencyCache      map[string][]float64
	invocationProbabilityCache map[string]float64
}

func NewInputManager(
	RegionViabilityLoader *loaders.RegionViabilityLoader,
	DataCenterLoader *loaders.DataCenterLoader,
	PerformanceLoader *loaders.PerformanceLoader,
	CarbonLoader *loaders.CarbonLoader,
	WorkflowLoader *loaders.WorkflowLoader,
	RuntimeCalculator *calculators.RuntimeCalculator,
	CarbonCalculator *calculators.CarbonCalculator,
	CostCalculator *calculators.CostCalculator,
	InstanceIndexer *models.InstanceIndexer,
	RegionIndexer *models.RegionIndexer,
) (*InputManager, error) {
	return &InputManager{
		RegionViabilityLoader:      RegionViabilityLoader,
		DataCenterLoader:           DataCenterLoader,
		PerformanceLoader:          PerformanceLoader,
		CarbonLoader:               CarbonLoader,
		WorkflowLoader:             WorkflowLoader,
		RuntimeCalculator:          RuntimeCalculator,
		CarbonCalculator:           CarbonCalculator,
		CostCalculator:             CostCalculator,
		InstanceIndexer:            InstanceIndexer,
		RegionIndexer:              RegionIndexer,
		executionLatencyCache:      map[string][]float64{},
		invocationProbabilityCache: map[string]float64{},
	}, nil
}

func Setup(data map[string]interface{}) *InputManager {
	regionViabilityLoader, _ := loaders.NewRegionViabilityLoader(data["region_viability_loader"].([]interface{}))

	datacenterLoader, _ := loaders.NewDataCenterLoader(
		data["datacenter_loader"].(map[string]interface{}),
	)

	performanceLoader, _ := loaders.NewPerformanceLoader(
		data["performance_loader"].(map[string]interface{}),
	)

	carbonLoader, _ := loaders.NewCarbonLoader(
		data["carbon_loader"].(map[string]interface{}),
	)

	workflowLoader, _ := loaders.NewWorkflowLoader(
		data["workflow_loader"].(map[string]interface{})["workflow_data"].(map[string]interface{}),
		data["workflow_loader"].(map[string]interface{})["instances_regions_and_providers"].(map[string]interface{}),
		data["workflow_loader"].(map[string]interface{})["home_region"].(string),
		performanceLoader,
	)

	runtimeCalculator, _ := calculators.NewRuntimeCalculator(
		performanceLoader, workflowLoader,
	)

	carbonCalculator, _ := calculators.NewCarbonCalculator(
		carbonLoader,
		datacenterLoader,
		workflowLoader,
		data["consider_cfe"].(bool),
		data["energy_factor"].(float64),
		data["carbon_free_intra_region_transmission"].(bool),
		data["carbon_free_dt_during_execution_at_home_region"].(bool),
	)

	costCalculator, _ := calculators.NewCostCalculator(
		datacenterLoader, workflowLoader, runtimeCalculator,
	)

	instanceIndexer := models.CopyInstanceIndexer(data["instance_indexer"].(map[string]interface{}))
	regionIndexer := models.CopyRegionIndexer(data["region_indexer"].(map[string]interface{}))

	inputManager, _ := NewInputManager(
		regionViabilityLoader,
		datacenterLoader,
		performanceLoader,
		carbonLoader,
		workflowLoader,
		runtimeCalculator,
		carbonCalculator,
		costCalculator,
		instanceIndexer,
		regionIndexer,
	)

	fmt.Println("Go: Created Input Manager")

	return inputManager
}

func (im *InputManager) AlterCarbonSetting(setting *string) {
	im.CarbonCalculator.AlterCarbonSetting(setting)
	im.RuntimeCalculator.ResetCache()

	// Clear the cache
	im.executionLatencyCache = map[string][]float64{}
	im.invocationProbabilityCache = map[string]float64{}
}

func (im *InputManager) GetInvocationProbability(fromInstanceIndex int, toInstanceIndex int) float64 {
	// 	Return the probability of the edge being triggered.
	// Check if the value is already in the cache
	key := strconv.Itoa(fromInstanceIndex) + "_" + strconv.Itoa(toInstanceIndex)
	if val, exists := im.invocationProbabilityCache[key]; exists {
		return val
	}

	// Convert the instance indices to their names
	fromInstanceName := im.InstanceIndexer.IndexToValue(fromInstanceIndex)
	toInstanceName := im.InstanceIndexer.IndexToValue(toInstanceIndex)

	// If not, retrieve the value from the workflow loader
	invocationProbability := im.WorkflowLoader.GetInvocationProbability(fromInstanceName, toInstanceName)
	im.invocationProbabilityCache[key] = invocationProbability
	return invocationProbability
}

func (im *InputManager) GetStartHopRetrieveWpdProbability() float64 {
	// Return the probability of workflow placement decision being retrieved at
	// the first function (or redirector) rather than the client CLI.
	// If not, retrieve the value from the workflow loader
	return im.WorkflowLoader.GetStartHopRetrieveWpdProbability()
}

func (im *InputManager) GetAllRegions() []string {
	return im.RegionViabilityLoader.AvailableRegions
}

func (im *InputManager) GetTransmissionInfo(
	fromInstanceIndex int,
	fromRegionIndex int,
	toInstanceIndex int,
	toRegionIndex int,
	cumulativeRuntime float64,
	toInstanceIsSyncNode bool,
	considerFromClientLatency bool,
) map[string]interface{} {
	// Convert the instance and region indices to their names
	// For start hop, from_instance_index and from_region_index will be -1
	fromInstanceName := ""
	fromRegionName := ""
	if fromInstanceIndex != -1 {
		fromInstanceName = im.InstanceIndexer.IndexToValue(fromInstanceIndex)
	}
	if fromRegionIndex != -1 {
		fromRegionName = im.RegionIndexer.IndexToValue(fromRegionIndex)
	}
	// To instance and region will always have region and instance index
	toInstanceName := im.InstanceIndexer.IndexToValue(toInstanceIndex)
	toRegionName := im.RegionIndexer.IndexToValue(toRegionIndex)
	// Get a transmission size and latency sample
	transmissionSize, transmissionLatency := im.RuntimeCalculator.CalculateTransmissionSizeLatency(
		fromInstanceName, fromRegionName, toInstanceName, toRegionName, toInstanceIsSyncNode, considerFromClientLatency,
	)
	snsTransmissionSize := transmissionSize
	var syncInfo map[string]interface{}
	if toInstanceIsSyncNode {
		// If to instance is a sync node, then at the same time,
		// we can retrieve the sync_sizes_gb and sns_only_sizes_gb
		// And then calculate the sync node related information.
		if len(fromInstanceName) == 0 {
			log.Fatalf("Start hop cannot have a sync node as a successor")
		}
		snsOnlySize, syncSize, wcu := im.GetUploadSyncSizeAndWcu(fromInstanceName, toInstanceName)
		snsTransmissionSize = snsOnlySize
		syncInfo = map[string]interface{}{
			"dynamodb_upload_size":                   transmissionSize,
			"sync_size":                              syncSize,
			"consumed_dynamodb_write_capacity_units": wcu,
			"sync_upload_auxiliary_info":             []float64{cumulativeRuntime, transmissionSize},
		}
	}
	// If to instance is a sync node, then at the same time,
	// // we can retrieve the sync_sizes_gb and sns_only_sizes_gb
	// And then calculate the sync node related information.
	return map[string]interface{}{
		"starting_runtime":       cumulativeRuntime,
		"cumulative_runtime":     cumulativeRuntime + transmissionLatency,
		"sns_data_transfer_size": snsTransmissionSize,
		"sync_info":              syncInfo,
	}
}

func (im *InputManager) GetUploadSyncSizeAndWcu(
	fromInstanceName string, toInstanceName string,
) (float64, float64, float64) {
	// If to instance is a sync node, then at the same time,
	// we can retrieve the sync_sizes_gb and sns_only_sizes_gb
	snsOnlySize := im.WorkflowLoader.GetSNSOnlySize(fromInstanceName, toInstanceName)
	syncSize := im.WorkflowLoader.GetSyncSize(fromInstanceName, toInstanceName)

	// We have to get sync_size * 2, as our wrapper does 2 update operations
	dynamodbWriteCapacityUnits := im.CalculateWriteCapacity(syncSize) * 2

	return snsOnlySize, syncSize, dynamodbWriteCapacityUnits
}

func (im *InputManager) CalculateWriteCapacity(dataSizeGb float64) float64 {
	// We can calculate the write capacity units for the data size
	// DynamoDB charges 1 WCU for 1 KB of data written for On-Demand capacity mode
	// https://aws.amazon.com/dynamodb/pricing/on-demand/

	// Convert the data size from GB to KB
	// And then round up to the nearest 1 KB
	dataSizeKb := math.Ceil(dataSizeGb * math.Pow(1024, 2))
	writeCapacityUnits := dataSizeKb

	return writeCapacityUnits
}

func (im *InputManager) calculateReadCapacityUnits(dataSizeGb float64) float64 {
	// We can calculate the read capacity units for the data size
	// DynamoDB charges 1 RCU for up to 4 KB of data read for On-Demand capacity mode
	// For strongly consistent reads (What our wrapper uses)
	// https://aws.amazon.com/dynamodb/pricing/on-demand/

	// Convert the data size from GB to KB
	// And then round up to the nearest 4 KB
	dataSizeKb := dataSizeGb * 1024 * 1024
	readCapacityUnits := math.Ceil(dataSizeKb / 4)

	return readCapacityUnits
}

func (im *InputManager) GetSimulatedTransmissionInfo(
	fromInstanceIndex int,
	uninvokedInstanceIndex int,
	simulatedSyncPredecessorIndex int,
	syncNodeIndex int,
	fromRegionIndex int,
	toRegionIndex int,
	cumulativeRuntime float64,
) map[string]interface{} {
	// Convert the instance and region indices to their names
	fromInstanceName := im.InstanceIndexer.IndexToValue(fromInstanceIndex)
	uninvokedInstanceName := im.InstanceIndexer.IndexToValue(uninvokedInstanceIndex)
	simulatedSyncPredecessorName := im.InstanceIndexer.IndexToValue(simulatedSyncPredecessorIndex)
	syncNodeName := im.InstanceIndexer.IndexToValue(syncNodeIndex)
	fromRegionName := im.RegionIndexer.IndexToValue(fromRegionIndex)
	toRegionName := im.RegionIndexer.IndexToValue(toRegionIndex)

	snsTransmissionSize, transmissionLatency := im.RuntimeCalculator.CalculateSimulatedTransmissionSizeLatency(
		fromInstanceName,
		uninvokedInstanceName,
		simulatedSyncPredecessorName,
		syncNodeName,
		fromRegionName,
		toRegionName,
	)
	return map[string]interface{}{
		"starting_runtime":       cumulativeRuntime,
		"cumulative_runtime":     cumulativeRuntime + transmissionLatency,
		"sns_data_transfer_size": snsTransmissionSize,
	}
}

func (im *InputManager) GetNonExecutionInfo(
	fromInstanceIndex int,
	toInstanceIndex int,
) map[string]interface{} {
	// Convert the instance and region indices to their names
	// Start hop will never get non-execution info
	fromInstanceName := im.InstanceIndexer.IndexToValue(fromInstanceIndex)
	toInstanceName := im.InstanceIndexer.IndexToValue(toInstanceIndex)

	nonExecutionInfoList := []map[string]interface{}{}
	for syncToFromInstance, syncSize := range im.WorkflowLoader.GetNonExecutionInformation(
		fromInstanceName, toInstanceName,
	) {
		parsedSyncToFromInstance := strings.Split(syncToFromInstance, ">")
		syncPredecessorInstance := parsedSyncToFromInstance[0]
		syncNodeInstance := parsedSyncToFromInstance[1]

		nonExecutionInfoList = append(
			nonExecutionInfoList,
			map[string]interface{}{
				"predecessor_instance_id":                im.InstanceIndexer.ValueToIndex(syncPredecessorInstance),
				"sync_node_instance_id":                  im.InstanceIndexer.ValueToIndex(syncNodeInstance),
				"sync_size":                              syncSize,
				"consumed_dynamodb_write_capacity_units": 2 * im.CalculateWriteCapacity(syncSize), // We have to get sync_size * 2, as our wrapper does 2 update operations
			},
		)
	}

	return map[string]interface{}{"non_execution_info": nonExecutionInfoList}
}

func (im *InputManager) GetNodeRuntimeDataTransfer(
	instanceIndex int, regionIndex int, previousCumulativeRuntime float64, isRedirector bool,
) (map[string]interface{}, float64, float64) {
	// Convert the instance and region indices to their names
	instanceName := im.InstanceIndexer.IndexToValue(instanceIndex)
	regionName := im.RegionIndexer.IndexToValue(regionIndex)

	// Get the node runtimes and data transfer information
	return im.RuntimeCalculator.CalculateNodeRuntimeDataTransfer(
		instanceName, regionName, previousCumulativeRuntime, im.InstanceIndexer, isRedirector,
	)
}

func (im *InputManager) CalculateCostCarbonOfInstance(
	executionTime float64,
	instanceIndex int,
	regionIndex int,
	dataInputSizes map[int]float64,
	dataOutputSizes map[int]float64,
	snsDataCallAndOutputSizes map[int][]float64,
	dataTransferDuringExecution float64,
	dynamodbReadCapacity float64,
	dynamodbWriteCapacity float64,
	isInvoked bool,
	isRedirector bool,
) map[string]float64 {
	// Convert the instance and region indices to their names
	instanceName := im.InstanceIndexer.IndexToValue(instanceIndex)
	regionName := im.RegionIndexer.IndexToValue(regionIndex)

	dataOutputSizesStrDict := GetConvertedRegionNameMap(im.RegionIndexer, dataOutputSizes)
	executionCarbon, transmissionCarbon := im.CarbonCalculator.CalculateInstanceCarbon(
		executionTime,
		instanceName,
		regionName,
		GetConvertedRegionNameMap(im.RegionIndexer, dataInputSizes),
		dataOutputSizesStrDict,
		dataTransferDuringExecution,
		isInvoked,
		isRedirector,
	)
	return map[string]float64{
		"cost": im.CostCalculator.CalculateInstanceCost(
			executionTime,
			instanceName,
			regionName,
			dataOutputSizesStrDict,
			GetConvertedRegionNameMap(im.RegionIndexer, snsDataCallAndOutputSizes),
			dynamodbReadCapacity,
			dynamodbWriteCapacity,
			isInvoked,
		),
		"execution_carbon":    executionCarbon,
		"transmission_carbon": transmissionCarbon,
	}
}

func (im *InputManager) CalculateCostCarbonVirtualStartInstance(
	dataInputSizes map[int]float64,
	dataOutputSizes map[int]float64,
	snsDataCallAndOutputSizes map[int][]float64,
	dynamodbReadCapacity float64,
	dynamodbWriteCapacity float64,
) map[string]float64 {
	dataOutputSizesStrDict := GetConvertedRegionNameMap(im.RegionIndexer, dataOutputSizes)

	return map[string]float64{
		"cost": im.CostCalculator.CalculateVirtualStartInstanceCost(
			dataOutputSizesStrDict,
			GetConvertedRegionNameMap(im.RegionIndexer, snsDataCallAndOutputSizes),
			dynamodbReadCapacity,
			dynamodbWriteCapacity,
		),
		"execution_carbon": 0.0,
		"transmission_carbon": im.CarbonCalculator.CalculateVirtualStartInstanceCarbon(
			GetConvertedRegionNameMap(im.RegionIndexer, dataInputSizes), dataOutputSizesStrDict,
		),
	}
}

func GetConvertedRegionNameMap[V any](regionIndexer *models.RegionIndexer, inputRegionIndexDict map[int]V) map[string]V {
	systemRegionFullName := fmt.Sprintf("aws:%s", calculators.GlobalSystemRegion)
	result := map[string]V{}
	for regionIndex, value := range inputRegionIndexDict {
		if regionIndex != -1 {
			result[regionIndexer.IndexToValue(regionIndex)] = value
		} else {
			if regionIndex == -2 {
				result[systemRegionFullName] = value
			} else {
				result[""] = value
			}
		}
	}
	return result
}

func (im *InputManager) CalculateDynamoDbCapacityUnitOfSyncEdges(
	syncEdgeUploadEdgesAuxiliaryData [][]float64,
) map[string]float64 {
	// Each entry of the sync_edge_upload_edges_auxiliary_data is a tuple
	// Where the first element is when a node reaches the invoke_call
	// Where the second element is the size of the sync uploads
	// The third element is the latency of the entire tranmsission (May be used)
	// We need to first sort the list by the first element (shortest time first)
	// Then we calculate the WRU for each entry with cumulative data sizes
	writeCapacityUnits := 0.0
	cumulativeDataSize := 0.0

	// Sort the list by the first element
	sort.Slice(
		syncEdgeUploadEdgesAuxiliaryData,
		func(i, j int) bool {
			return syncEdgeUploadEdgesAuxiliaryData[i][0] < syncEdgeUploadEdgesAuxiliaryData[i][1]
		},
	)
	for _, entry := range syncEdgeUploadEdgesAuxiliaryData {
		// The entry is a tuple of (cumulative_runtime, sync_size, transmission_latency)
		syncSize := entry[1]
		cumulativeDataSize += syncSize
		writeCapacityUnits += im.CalculateWriteCapacity(cumulativeDataSize)
	}
	return map[string]float64{
		"read_capacity_units":  im.calculateReadCapacityUnits(cumulativeDataSize),
		"write_capacity_units": writeCapacityUnits,
	}
}

func (im *InputManager) GetStartHopInfo() map[string]float64 {
	workflowPlacementDecisionSizeGb := im.WorkflowLoader.GetWorkflowPlacementDecisionSize()
	readCapacityUnits := im.calculateReadCapacityUnits(workflowPlacementDecisionSizeGb)
	return map[string]float64{
		"read_capacity_units":              readCapacityUnits,
		"workflow_placement_decision_size": workflowPlacementDecisionSizeGb,
	}
}

func (im *InputManager) GetHomeRegionIndex() int {
	return im.RegionIndexer.ValueToIndex(im.WorkflowLoader.HomeRegion)
}

func (im *InputManager) TestFunc1(dataString string) float64 {
	return 1.1
}

func (im *InputManager) TestFunc2(dataString string) []float64 {
	return []float64{1.1, 1.2, 1.3}
}
